coherence-data/src/main/java/io/micronaut/coherence/data/ops/DefaultCoherenceAsyncRepositoryOperations.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.coherence.data.ops;
+
+import com.tangosol.coherence.dslquery.ExecutionContext;
+import com.tangosol.coherence.dslquery.Statement;
+import com.tangosol.coherence.dslquery.StatementResult;
+import com.tangosol.net.AsyncNamedMap;
+import com.tangosol.util.Processors;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.micronaut.coherence.data.annotation.AsyncPersistEventSource;
+import io.micronaut.coherence.data.annotation.AsyncRemoveEventSource;
+import io.micronaut.coherence.data.annotation.AsyncUpdateEventSource;
+import io.micronaut.context.annotation.Parameter;
+import io.micronaut.context.annotation.Prototype;
+import io.micronaut.data.exceptions.EmptyResultException;
+import io.micronaut.data.model.Page;
+import io.micronaut.data.model.runtime.DeleteBatchOperation;
+import io.micronaut.data.model.runtime.DeleteOperation;
+import io.micronaut.data.model.runtime.InsertBatchOperation;
+import io.micronaut.data.model.runtime.InsertOperation;
+import io.micronaut.data.model.runtime.PagedQuery;
+import io.micronaut.data.model.runtime.PreparedQuery;
+import io.micronaut.data.model.runtime.UpdateOperation;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+/**
+ * Concrete {@link CoherenceAsyncRepositoryOperations} implementation using
+ * a {@code Coherence} {@link AsyncNamedMap}.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+@Prototype
+class DefaultCoherenceAsyncRepositoryOperations
```

Comment:
I'm actually ok with `thenApply` usage here instead of `thenApplyAsync`, as we control the code that gets executed and everything we do seems too light-weight to warrant thread switch. So, leave it as is.